sdks/go/pkg/beam/metrics.go

```diff
@@ -0,0 +1,373 @@
+package beam
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/apache/beam/sdks/go/pkg/beam/core/util/ptransform"
+	"github.com/apache/beam/sdks/go/pkg/beam/log"
+	"github.com/apache/beam/sdks/go/pkg/beam/model/fnexecution_v1"
+	"github.com/golang/protobuf/ptypes"
+)
+
+type userMetricker interface {
+	toProto() *fnexecution_v1.Metrics_User
+}
+
+type metricName struct {
+	namespace, name string
+}
+
+func (n *metricName) String() string {
+	return fmt.Sprintf("%s.%s", n.namespace, n.name)
+}
+
+func validateName(mn metricName, t metricType) {
+	if len(mn.name) == 0 || len(mn.namespace) == 0 {
+		panic(fmt.Sprintf("namespace and name are required to be non-empty, got %q and %q", mn.namespace, mn.name))
+	}
+}
+
+var (
+	metricMu      sync.Mutex
+	metricStorage = make(map[string]map[metricName]userMetricker)
+	counters      = make(map[metricName]*Counter)
+	distributions = make(map[metricName]*Distribution)
+	gauges        = make(map[metricName]*Gauge)
+)
+
+type metricType uint8
+
+const (
+	counterType metricType = iota
+	distributionType
+	gaugeType
+)
+
+func (t metricType) String() string {
+	switch t {
+	case counterType:
+		return "Counter"
+	case distributionType:
+		return "Distribution"
+	case gaugeType:
+		return "Gauge"
+	default:
+		panic(fmt.Sprintf("Unknown metric type value: %v", int(t)))
```

Comment:
Shouldn't need the cast here, %v knows how to render a uint8.

Reply:
Done, where appropriate.